### PR TITLE
First CPU implementation of HashedNets compression algorithm for DNNs

### PR DESCRIPTION
**Pull request by adi-sharma**
This is an implementation of HashedNets (https://arxiv.org/pdf/1504.04788.pdf) in ROOT.

It is built over the existing implementation of DNNs in TMVA and supports multithreading.

### CODE REVIEW

File: tmva/tmva/src/MethodDNN.cxx [hunk 1]
```diff
@@ -1100,9 +1102,9 @@ void TMVA::MethodDNN::checkGradients ()
    fNet.addLayer (DNN::Layer (outputSize, DNN::EnumFunction::LINEAR, DNN::ModeOutputValues::SIGMOID)); 
    fNet.setErrorFunction (DNN::ModeErrorFunction::CROSSENTROPY);
    //    net.setErrorFunction (ModeErrorFunction::SUMOFSQUARES);
-
+   const int BUCKET_SIZE = 8;
```
Comment (omazapa) [inline tmva/tmva/src/MethodDNN.cxx]: Hi Aditya,

Why bucket size is  constant value here?

Comment (adi-sharma) [inline tmva/tmva/src/MethodDNN.cxx]: Hi Omar, that's because this is the checkGradients () function. As you can see above at lines 1100 - 03, this is a predefined Neural Net created just to see if the algorithm calculates gradients properly. This is independent of the actual training.

**Comment by vgvassilev** [comment]
Could you rebase this PR and squash the duplicate commits?

**Comment by adi-sharma** [comment]
This patch needs some work. I'm closing the pull request for now. Will issue a new one after resolving the issues.
